pkg/escapism/escapism.go

```diff
@@ -0,0 +1,31 @@
+/*
+Full functions, please reference:
+https://github.com/minrk/escapism
+*/
+
+package escapism
+
+import (
+	"encoding/hex"
+	"unicode"
+)
+
+func escapeChar(c string, escape_char string) string {
+	bytechars := []byte(c)
+	var safechars string
+	safechars += escape_char
+	safechars += hex.EncodeToString(bytechars)
+	return safechars
+}
+
+func Escape(input string) string {
+	safestring := ""
+	for _, char := range input {
+		if !unicode.IsLower(char) && !unicode.IsDigit(char) {
```

Comment:
According to k8s spec label name can contain `[a-z0-9A-Z]`, `-`, `_` and `.`
https://kubernetes.io/docs/concepts/overview/working-with-objects/labels/

It looks like here we only allow lower case and digits, do we really need to escape those in the whitelists? What will the encoded string like if the original string contains Uppercase, `_`, ``- and `.`

Reply:
Follow the convention from `kubespawner`:
https://github.com/jupyterhub/kubespawner/blob/1b4539c114d6aa46232553dfd1ac4c48295aba2b/kubespawner/spawner.py#L1307

It will convert to hex and add prefix `-` if it's not a safe char.
EX: "A" => "-41"

Do you think we can follow this convention?

BTW, I change PR to WIP since I found somethings need to be modified.

Reply:
If we don't have strong reason then I would like to keep those characters, because it's more human readable.

Say I have an email `ash@InfuseAI.com`, it would be converted to `ash-40-49nfuse-41-49-20com`, which looks weird to me and I recognize it at a glance.

Reply:
Let's get some feedbacks from @clkao ~ He recommends following how kubespawner deals with this in the beginning.

Reply:
Sure, it's also a good reason - keep the same encoding logic across components.

Reply:
On second thought, I think it's better for now to use the same encoding logic in ph-jobs and jupyterhub. Although I still have doubt about it. Like: how can we map back the original email? Is the decoding ambiguous?

But as for fixing the issue itself, I think it's ok to leave it that way.

Reply:
Yes, thanks. My current implementation is not totally the same as python's version. Maybe cannot be convert back. Let me check and fix and add tests.

Reply:
@hSATAC it's one to one mapping now. Also add unescape function and it's unit test. thx.

Reply:
Cool! Thanks! LGTM 💯